Default choice selection to empty when generating new api

jingcheng16 opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->
**Before**
We used to default api key's scope to all projects this web user has access to
<img width="523" height="318" alt="image" src="https://github.com/user-attachments/assets/3064c239-6b12-4641-abf8-2745e5c52b23" />

**After**
We now will default to no selection, and move the option "All Projects" to the bottom of the list so we discourage user from creating unscoped key
<img width="846" height="620" alt="image" src="https://github.com/user-attachments/assets/d9519106-51d9-4ea5-832c-818dd9d54b3e" />

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
Ticket: https://dimagi.atlassian.net/browse/SAAS-18096

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
I've tested locally. This is just change what options we provided and the order of it.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

corehq.apps.settings.tests.test_forms::HQApiKeyTests

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->


### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

jingcheng16 committed:
default selection to empty and move "All Projects" to the bottom of the list

nospame reviewed (commented):
This code seems functional as written, but having to change the value during form validation feels weird.

jingcheng16 committed:
Move the logic to handle "ALL_DOMAINS" to the clean_field method

jingcheng16 committed:
Update the test

I didn't add a test to make sure a form without domain is invalid. I feel this is a redundancy to test a required field. and some old test already failed due to domain is not included in the `_form_data` by default

gherceg reviewed (commented):
I don't love the default selection being a blank value. Is there not just a way to default to no selection?

mjriley reviewed (changes requested):
Changes look good, but I'd like to see a test added to verify the new behavior

kaapstorm commented on corehq/apps/settings/forms.py:
```diff
@@ -292,7 +292,8 @@ def __init__(self, *args, user_domains=None, max_allowed_expiration_days=None, t
         user_domains = user_domains or []
         all_domains = (self.ALL_DOMAINS, _('All Projects'))
-        self.fields['domain'].choices = [all_domains] + [(d, d) for d in user_domains]
+        default_empty_choice = ('', '')
```
Should we be more explicit that this option is not going to work out?
```suggestion
        default_empty_choice = ('', _('(None)'))
```

kaapstorm commented on corehq/apps/settings/forms.py:
```diff
@@ -269,7 +269,7 @@ def __init__(self, **kwargs):
 
 class HQApiKeyForm(forms.Form):
-    ALL_DOMAINS = ''
+    ALL_DOMAINS = 'ALL_DOMAINS'  # This value is safe to use because we normalize all domain names
```
Instead of creating a new value, which would require a migration to preserve current behavior (or break all API keys that users originally intended to be used for all domains), should we rather create a new value for "nothing selected". Maybe something like ...
```python
    NO_DOMAINS = 'NO_DOMAINS'
    # ...
    no_domains = (self.NO_DOMAINS, _('(None)'))
    self.fields['domain'].choices = [no_domains] + [(d, d) for d in user_domains] + [all_domains]
```

mjriley replied:
I believe the issue with that approach is that `NO_DOMAINS` shouldn't be a legal value, right? This PR changed the domain field to be a required field, with the benefit being that the user gets front-end feedback that the empty value isn't allowed. If we were to adopt a sentinel value for no domains, we'd either need to introduce separate javascript functionality to alert the user of the error before submission, or have the clunky user experience where we allow them to submit a bad form.

The code here doesn't require a migration, because it ultimately cleans `ALL_DOMAINS` to `''` prior to creating the api key. At least within the context of `HQApiKey`, `''` still means "all domains", so no existing keys get broken. However, I believe Evan and myself were concerned that manipulating the value like this isn't particularly intuitive. I don't see a great way to handle this without creating a migration and updating a bunch of other code, which seems out of scope for this kind of change. I don't think that's what Kai had in mind when requesting this.

kaapstorm commented on corehq/apps/settings/forms.py:
```diff
@@ -345,19 +345,20 @@ def create_key(self, user):
             return new_key
 
+    def clean_domain(self):
+        if self.cleaned_data['domain'] == self.ALL_DOMAINS:
+            return ''
```
The bit that I missed on my first pass. :upside_down_face: 

> What do you think of changing the `ALL_DOMAINS` constant to something like `ALL_DOMAINS_UI` or `ALL_DOMAINS_RAW` or something

I second the idea of renaming this. Something like `ALL_DOMAINS_PLACEHOLDER` would have helped me understand its purpose.

jingcheng16 replied:
Thank you both! Renamed in 3d216fde13af9381166166cf2000f9f68b4c32b4

jingcheng16 committed:
Update variables

jingcheng16 committed:
Add new test

jingcheng16 committed:
Explicitly set domain to ALL_DOMAINS_UI and only test against the clean_field method

mjriley reviewed (approved):
Changes look good. My only remaining concern is that someone still might be confused why we need to change `ALL_DOMAINS_UI` to `ALL_DOMAINS` during the cleaning process, so the code might benefit from a comment explaining that. But I don't think that blocks this PR
